libs/ts-rest/core/src/lib/error-status-error.ts

```diff
@@ -0,0 +1,54 @@
+import { AppRoute } from './dsl';
+import { ErrorResponse } from './infer-types';
+import { ErrorHttpStatusCode } from './status-codes';
+
+export class ErrorStatusResponseError<
+  TAppRoute extends AppRoute,
+  TErrorStatus extends ErrorHttpStatusCode = ErrorHttpStatusCode,
+> extends Error {
+  constructor(
+    public readonly route: TAppRoute,
+    public readonly response: ErrorResponse<TAppRoute, TErrorStatus>,
+    public readonly request: unknown,
+  ) {
+    super(`Request failed with status ${response.status}`);
+  }
+}
+
+/**
+ * Check if the given `error` is an instance of {@link ErrorStatusResponseError}
+ * and corresponds to the given `route` and matches the given `status`.
+ *
+ * @param route The route to check against
+ * @param status The status to check against.
+ *
+ * - For single status it narrows down to a single response.
+ * - Array of status codes it creates a union type of all the responses that match the status codes.
+ * - If no given status, it will narrow to all error responses in the route.
+ */
+export function isErrorStatusResponse<
+  TAppRoute extends AppRoute,
+  TErrorStatus extends ErrorHttpStatusCode = ErrorHttpStatusCode,
+>(
+  error: unknown,
+  route: TAppRoute,
+  status?: TErrorStatus | TErrorStatus[],
+): error is ErrorStatusResponseError<
+  TAppRoute,
+  TErrorStatus extends undefined ? ErrorHttpStatusCode : TErrorStatus
```

Comment:
`TErrorStatus extends undefined` will never be true because you already declared it as `TErrorStatus extends ErrorHttpStatusCode`. Even if not, `T extends undefined` in general doesn't work that way for optional parameters, because they resolve as `unknown` not `undefined`, and actually because `extends` is distributive you need to do `unknown extends T` not `T extends unknown`

This should work fine if you just replace `TErrorStatus extends undefined ? ErrorHttpStatusCode : TErrorStatus` with `TErrorStatus` because you already gave it a default value of `ErrorHttpStatusCode`, so it will take that if you don't pass `status`